### PR TITLE
[FLINK-13119] add blink table config to documentation

### PR DESCRIPTION
**Pull request by XuPingyong**
## What is the purpose of the change

1. Add table config annotation to indicate it works for batch or streaming jobs.
2. Add table config to documentation.

The preview screenshot of configuration page looks like:

image

## Verifying this change

*(Please pick either of the following options)*

no

This change added tests and can be verified as follows:

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? ( no)

### CODE REVIEW

**Comment by XuPingyong** [comment]
@zentol,  https://github.com/apache/flink/pull/8937 mixes multiple unrelated changes, and I split it to 2 prs:
1. https://github.com/apache/flink/pull/8980 (merged)
2. this pr.

File: docs/_includes/generated/execution_config_configuration.html [hunk 1]
```diff
@@ -0,0 +1,122 @@
+<table class="table table-bordered">
```
Comment (twalthr) [inline docs/_includes/generated/execution_config_configuration.html]: This config does not belong into the Flink core config documentation. But into a dedicated Table/Configuration page

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java [hunk 2]
```diff
@@ -260,10 +262,26 @@ private static String toHtmlTable(final List<OptionWithMetaInfo> options) {
 	private static String toHtmlString(final OptionWithMetaInfo optionWithMetaInfo) {
 		ConfigOption<?> option = optionWithMetaInfo.option;
 		String defaultValue = stringifyDefault(optionWithMetaInfo);
+		Documentation.TableMeta tableMeta = optionWithMetaInfo.field.getAnnotation(Documentation.TableMeta.class);
+		StringBuilder sb = new StringBuilder();
+		if (tableMeta != null) {
+			Documentation.ExecMode execMode = tableMeta.execMode();
+			if (Documentation.ExecMode.BATCH_STREAMING.equals(execMode)) {
+				sb.append("<br> <span class=\"label label-primary\">")
+						.append("BATCH")
```
Comment (wuchong) [inline flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Use `ExecMode.BATCH` instead.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java [hunk 3]
```diff
@@ -260,10 +262,26 @@ private static String toHtmlTable(final List<OptionWithMetaInfo> options) {
...
+						.append("</span> <span class=\"label label-primary\">")
+						.append("STREAMING")
```
Comment (wuchong) [inline flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Use `Exec.STREAMING` instead.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java [hunk 4]
```diff
@@ -260,10 +262,26 @@ private static String toHtmlTable(final List<OptionWithMetaInfo> options) {
 	private static String toHtmlString(final OptionWithMetaInfo optionWithMetaInfo) {
 		ConfigOption<?> option = optionWithMetaInfo.option;
 		String defaultValue = stringifyDefault(optionWithMetaInfo);
+		Documentation.TableMeta tableMeta = optionWithMetaInfo.field.getAnnotation(Documentation.TableMeta.class);
+		StringBuilder sb = new StringBuilder();
```
Comment (wuchong) [inline flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Use a meaningful local field name? For example: `tags`, `metaStr`

File: docs/_includes/generated/execution_config_configuration.html (hunk 1 above)
Comment (wuchong) [reply docs/_includes/generated/execution_config_configuration.html]: Agree. We should have a separate page, for example: `docs/dev/table/config.md` and add some more description on the top of the page.

**Comment by wuchong** [comment]
Hi @zentol do you want to have a look again? 
I think the `TableMeta` annotation is great to generate a good SQL configuration page.

File: docs/_includes/generated/execution_config_configuration.html (hunk 1 above)
Comment (XuPingyong) [reply docs/_includes/generated/execution_config_configuration.html]: Thanks @twalthr  @wuchong , I have move this to a dedicated page.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java (hunk 3 above)
Comment (XuPingyong) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: I think STREAMING is enough.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java (hunk 4 above)
Comment (XuPingyong) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Updated it.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java (hunk 2 above)
Comment (XuPingyong) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: I think BATCH is enough.

**Comment by wuchong** [comment]
I think we need to refactor the config page. I wrote a draft for this and pushed the commit to my repo: https://github.com/wuchong/flink/tree/pull/9024. 
Can you pick the commit to your branch @XuPingyong and continue the reviewing?

Here is a preview.
image

File: flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java [hunk 5]
```diff
@@ -177,7 +177,8 @@ private static void compareDocumentedAndExistingOptions(Map<String, DocumentedOp
 			.map(element -> element.getElementsByTag("tbody").get(0))
 			.flatMap(element -> element.getElementsByTag("tr").stream())
 			.map(tableRow -> {
-				String key = tableRow.child(0).text();
+				// split space to exclude document tags.
+				String key = tableRow.child(0).text().split(" ")[0];;
```
[the remark this answers is not in the document]
Comment (XuPingyong) [reply flink-docs/src/test/java/org/apache/flink/docs/configuration/ConfigOptionsDocsCompletenessITCase.java]: Exclude tag label.

**Comment by XuPingyong** [comment]
> I think we need to refactor the config page. I wrote a draft for this and pushed the commit to my repo: https://github.com/wuchong/flink/tree/pull/9024.
> Can you pick the commit to your branch @XuPingyong and continue the reviewing?
> 
> Here is a preview.
> image

Thanks @wuchong , I have merged it.

**Comment by wuchong** [comment]
Thanks for the updating. It looks good to me now.

@twalthr  do you want to have a look at the Configuration page?

**Comment by twalthr** [comment]
@godfreyhe @wuchong I would wait until the discussion in #9185 is done. The options should be follow the [Flink convention](https://docs.google.com/document/d/1JGRVK6udBFa8xbyv7ipPQiLEhxebA64f8KKcaAH0UPY/edit#heading=h.qmv83r3emtom). And I also wonder why it is prefixed with `sql.`, I would vote for `table.`. Otherwise we could also keep them in Blink for now.

**Comment by wuchong** [comment]
@twalthr Sure. It makes sense to continue this before we reach a consensus.

**Comment by wuchong** [comment]
Hi @XuPingyong , #9185 and #9203 is merged. I think we can resume this PR. 

Could you rebase and re-generate the config html?

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java (hunk 2 above)
Comment (zentol) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: I believe the point was to use `ExecMode.BATCH.toString()`, to ensure it is in sync with the `else` branch.

Comment (wuchong) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Yes.

**Comment by XuPingyong** [comment]
> Hi @XuPingyong , #9185 and #9203 is merged. I think we can resume this PR.
> 
> Could you rebase and re-generate the config html?

@wuchong  Thanks， done it.

File: docs/dev/table/blink.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+---
```
Comment (wuchong) [inline docs/dev/table/blink.md]: Please rename this file to `config.md`

File: docs/dev/table/blink.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+t_env.get_config()
+  .get_configuration()
+  .set_string("sql.exec.mini-batch.enabled", "true")
+  .set_string("sql.exec.mini-batch.allow-latency", "5 s")
+  .set_string("sql.exec.mini-batch.size", "5000");
+{% endhighlight %}
+</div>
+</div>
+
+**NOTE: currently, all the options are only works in blink planner. See [how to use blink planner](index.html).**
```
Comment (wuchong) [inline docs/dev/table/blink.md]: works -> supported

File: docs/dev/table/blink.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+// environment configuration
+TableEnvironment tEnv = ...
+
+// set key-value options
+tEnv.getConfig()
+  .getConfiguration()
+  .setString("sql.exec.mini-batch.enabled", "true")
```
Comment (wuchong) [inline docs/dev/table/blink.md]: Please change to the new config keys.

File: flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java (hunk 2 above)
Comment (wuchong) [reply flink-docs/src/main/java/org/apache/flink/docs/configuration/ConfigOptionsDocGenerator.java]: Hi @XuPingyong , could you update this too for a better protection?

File: docs/dev/table/blink.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
+---
+title: "Configuration"
+nav-id: "config"
+nav-parent_id: tableapi
+nav-pos: 4
```
Comment (wuchong) [inline docs/dev/table/blink.md]: Change the `nav-pos` to make config page to be the last one under "Table API & SQL"?

**Comment by twalthr** [comment]
@wuchong if this PR is in a good shape from your side. I would volunteer to merge it and add some more introductory words and fix a couple of typos if that is ok?

**Comment by wuchong** [comment]
Yes. It looks good from my side. Thanks for the help @twalthr .

**Pull request closed without merging**